# 2.문자열 자료형

## 2.1.문자열 사용

- Python의 문자열은 많은 편의성과 다양한 기능을 제공한다.
- Python은 문자열 데이터를 표현하기 위한 문법을 제공한다.

**1) 문자열 표현**

- **작은 따옴표**, 또는 **큰 따옴표**를 사용해 문자열을 표현한다.
- **len()** 함수로 문자열 길이(=포함된 문자 개수)를 확인할 수 있다.

In [2]:
# 큰 따옴표(" ")로 문자열 출력(Hello Python!)
str1 = "Hello Python!"
print(str1)

Hello Python!


In [11]:
# 작은 따옴표(' ')로 문자열 출력(Hello Python!)
str2 = 'Hello Python!'
print(str2)

Hello Python!


In [4]:
# 문자열 길이 확인
len(str1)

13

**2) 문자열 나열**

- 문자열을 콤마 없이 나란히 나열하면 합쳐진 결과를 얻는다.



In [5]:
str3 = 'Hello' 'Python'
str3

'HelloPython'

- 문자열을 콤마로 구분하여 나열하면 튜플(Tuple)이 된다.
- 참고로 튜플은 (1, 2 ,3) 형태로 소괄호 안에 요소가 나열된 자료형이다.

In [6]:
str4 = 'Hello', 'Python'
str4

('Hello', 'Python')

**3) 여러 줄 문자열 표현**

- 문자열 앞뒤에 따옴표 세 개를 나열하거나, \\n을 사용해 여러 줄 표현이 가능하다.
- print() 함수를 사용해야 의도한 내용이 제대로 표시된다.

In [7]:
str5 = "여러 줄의 문자열을 표시할 때는 \n을 사용하세요."
print(str5)

여러 줄의 문자열을 표시할 때는 
을 사용하세요.


In [9]:
str6 = """또 다른 방법은
큰 따옴표나 작은 따옴표를
세 개 사용하는 것입니다."""
print(str6)

또 다른 방법은
큰 따옴표나 작은 따옴표를 
세 개 사용하는 것입니다.


**4) 문자열에 따옴표 넣기**

- 큰 따옴표 안에 작은 따옴표, 또는 작은 따옴표 안에 큰 따옴표를 사용한다.
- \\' 또는 \\"를 사용해도 된다.
- print() 함수를 사용해야 의도한 내용이 제대로 표시된다.

In [13]:
str7 = '문자열 안에 따옴표, 예를 들어 "를 넣고 싶을 때는 이렇게 사용하세요.'
print(str7)

문자열 안에 따옴표, 예를 들어 "를 넣고 싶을 때는 이렇게 사용하세요.


In [23]:
str8 = "아니면 \' 또는 \"를 사용해도 됩니다."
print(str8)

아니면 ' 또는 "를 사용해도 됩니다.


5) 많이 사용되는 이스케이프(escape) 문자

- 이스케이프(escape) 코드란 프로그래밍할 때 사용할 수 있도록 미리 정의해 둔 '문자 조합'을 말한다.
- 주로 출력물을 보기 좋게 정렬하는 용도로 사용한다.

  * <b>\\"</b>: 큰따옴표를 출력한다.
  * <b>\\'</b>: 작은따옴표를 출력한다.
  * <b>\\n</b>: 줄바꿈(new line) 문자를 출력한다.
  * <b>\\t</b>: 탭(tab) 문자를 출력한다.
  * <b>\\\\</b>: 백슬래시(backslash) 문자를 출력한다.

In [24]:
print("호동: \"안녕하세요.\"")
print("재석: \'잘 지냈어요?\'")
print("Hello World!\nHello Python!")
print("홍길동\t27세\tgildong@gmail.com\t프로그래머")
print("백슬래시 출력: \\")

호동: "안녕하세요."
재석: '잘 지냈어요?'
Hello World!
Hello Python!
홍길동	27세	gildong@gmail.com	프로그래머
백슬래시 출력: \


## 2.2.문자열 연산

- 문자열 연산은 숫자 연산과 다르다.
- 문자열끼리 덧셈을 수행하면, 단순히 두 문자열이 연결된다.

**1) 문자열 + 문자열 = 문자열 결합**

In [25]:
s1 = 'hello'
s2 = 'world'

result = s1 + ' ' + s2
print(result)

hello world


**2) 문자열 * 정수 = 문자열 반복**

       *****************
       ** hello world **
       *****************

In [27]:
print('*'* 17)
print("** hello world **")
print('*'* 17)

*****************
** hello world **
*****************


In [28]:
# 문자열에 0을 곱하면 문자열이 제거됨
str ='hello'
print(str*1)
print(str*0)
print(str*2)

hello

hellohello


## 2.3.문자열 인덱싱과 슬라이싱

- 문자열은 문자 요소가 모여서 만들어진 자료형, 즉 **'ABC'는 'A', 'B', 'C' 세 요소가 모인 것**
- <font color="red">인덱스 값은 0부터 시작</font>
- 문자열 인덱싱과 슬라이싱은 이후에 다룰 리스트(List)와 비슷한 특성을 가진다.

In [31]:
str = '대한민국'
print(str)
print(str[0])
print(str[3])

대한민국
대
국


**1) 문자열 인덱싱**

- 인덱싱(Indexing): 정수형 위치 인덱스를 사용해 원하는 요소를 조회하는 것
- 첫 번째 문자는 인덱스 0에 해당한다.

In [32]:
s = '동해물과백두산이마르고'

print(s[1])  # 해
print(s[5])  # 두
print(s[-3])  # 마

해
두
마


**2) 문자열 슬라이싱**
- 슬라이싱(Slicing) : 범위를 지정해 일정 범위의 요소를 조회하는 것
- 슬라이싱은 보통 두 개의 인덱스로 구성되는데, **변수명[시작 인덱스:끝 인덱스]** 의 형태를 가진다.

In [40]:
s = '동해물과백두산이마르고'

print(s)
print(s[4:6])  # 백두
print(s[:4])   # 동해물과
print(s[-3:])   # 마르고
print(s[:])

동해물과백두산이마르고
백두
동해물과
마르고
동해물과백두산이마르고


**[참고] 인덱싱으로 문자열 변경 불가**

- 인덱싱, 슬라이싱을 사용해 문자를 변경할 수는 없다.
- 파이썬에서 문자열은 값을 변경할 수 없기 때문에, 불변(immutable) 객체라고도 한다.

In [41]:
str = '나의살던고향은꽃피는산골'
str[1] = '에'

TypeError: 'str' object does not support item assignment

**[참고] 슬라이싱의 범위**
- 문자열 인덱싱을 할 때는 범위를 벗어나는 경우 오류가 발생하지만, 슬라이싱의 경우 부드럽게 처리된다.

In [42]:
str1 = 'I love you'

print(str1[2:100000])

love you


## 2.4.문자열 포맷팅(Formatting)

- 문자열 일부분만 바꾸면서 해당 문자열을 재사용할 경우가 있다.
- 예를 들어 문자열에서 이름만 변경하며 사용하는 경우이다.
- 이를 **문자열 포맷팅**이라고 하며, **문자열.format()** 형태로 사용한다.


###2.4.1.% 포맷팅
- % 연산자를 이용하여 문자열에 변수나 값을 포함시킬 수 있다.

- 문자열 포맷 코드
    - %s : 문자열(String)
    - %c : 문자 1개(character)
    - %d : 정수(Integer)
    - %f : 부동소수(floating-point)

In [46]:
# 숫자 바로 대입
print('I eat %d mangos.' %9)

I eat 9 mangos.


In [47]:
# 문자열 바로 대입
print('I eat %s mangos.' %"six")

I eat six mangos.


In [48]:
# 2개 이상 값 넣기

print('I ate %d mangos. So I was scik for %s days.' %(20, "six"))

I ate 20 mangos. So I was scik for six days.


In [50]:
num = 20
day = "six"
print('I ate %d mangos. So I was scik for %s days.' %(num, day))

I ate 20 mangos. So I was scik for six days.


###2.4.2.문자열.format()
- 중괄호{}를 사용하여 변수를 표시하고 값을 전달할 수 있다.

**1) 기본 사용법**

- 단순 문자열에서 { }는 아무 의미가 없다.

In [51]:
# 단순 문자열
'이름:{}, 과목:{}, 점수:{}'

'이름:{}, 과목:{}, 점수:{}'

- 하지만 format() 메소드를 만나면 { }는 큰 의미를 가진다.

In [52]:
# format 사용
'이름:{}, 과목:{}, 점수:{}'.format('홍길동', '수학', 85)     # '이름:홍길동, 과목:수학, 점수:85'

'이름:홍길동, 과목:수학, 점수:85'

In [ ]:
# format 사용
print('이름:{0}, 과목:{1}, 점수:{2}'.format('홍길동', '수학', 85))     # '이름:홍길동, 과목:수학, 점수:85'

이름:홍길동, 과목:수학, 점수:85


- format() 메소드는 대부분 print() 함수 안에서 사용한다.

**2) 자릿수와 정렬 방식 지정**

- 좀 더 세밀한 출력 형태를 지정하고 싶은 경우 사용한다.

In [55]:
# 소수점 자리수 제한
print('{:.3f}'.format(12.1234567))

12.123


In [56]:
# 기본 정렬
print('[{0:10}][{1:6}]'.format('홍길동', 1200))

[홍길동       ][  1200]


In [57]:
# 오른쪽, 왼쪽 정렬
print('[{0:>10}][{1:<6}]'.format('홍길동', 1200))

[       홍길동][1200  ]


In [58]:
# 가운데 정렬
print('[{0:^10}][{1:^6}]'.format('홍길동', 1200))

[   홍길동    ][ 1200 ]


In [59]:
# 공백을 다른 문자로 채우기
print('[{0:-^10}][{1:_^6}]'.format('홍길동', 1200))

[---홍길동----][_1200_]


**<연습문제>**


[문1] 자신의 이름과 학창시절 국어, 수학, 영어 점수를 각 변수에 대입하시오.

In [60]:
# 변수 선언
name = "홍길동"
kor = 100
math = 90
eng = 80

[문2] 세 과목의 합을 구하시오.

In [61]:
# 합 구하기
my_sum = kor + math + eng
my_sum

270

[문3] 세 과목의 평균을 구하시오.

In [62]:
# 평균 구하기
my_mean = my_sum/3
my_mean

90.0

[문4] 다음 코드의 잘못 된 부분을 수정해서 결과가 제대로 표시되게 하시오.

In [ ]:
# 수정 전
print('{} 학생의 점수 합은 {} 이고 평균은 {:.2f} 입니다.'.format(Name, Mean, Sum))

NameError: name 'Name' is not defined

In [63]:
# 수정 후
print('{} 학생의 점수 합은 {} 이고 평균은 {:.2f} 입니다.'.format(name, my_sum, my_mean))

홍길동 학생의 점수 합은 270 이고 평균은 90.00 입니다.


### 2.4.3.f-String

- 파이썬 3.6 버전부터는 f-문자열 포맷팅 기능을 사용할 수 있다.
- 문자열 앞에 f 접두사를 붙이면 f-문자열 포맷팅 기능을 사용할 수 있다.

In [64]:
# 변수 선언
name = '홍길동'
age = 25
score = 2345.6789


In [65]:
print(f'이름:{name}')
print(f'나이:{age}')
print(f'점수:{score}')
print(f'{age}살의 {name}의 점수는 {score}입니다.')

이름:홍길동
나이:25
점수:2345.6789
25살의 홍길동의 점수는 2345.6789입니다.


- 출력되는 형태를 나름으로 조절할 수 있다. (참고: 문자열:s, 정수:d, 실수:f)

In [66]:
# 확인
print(f'{age}살의 {name}의 점수는 {score:,.2f}')

25살의 홍길동의 점수는 2,345.68


**<연습문제>**


[문1] 우선 다음 코드를 실행하여 점수 합계와 평균을 구하시오.

In [67]:
# 변수 선언
name = '홍길동'
age = 25
score1 = 153
score2 = 204
score3 = 259
score4 = 221
score5 = 267

# 합계
total_score = score1 + score2 + score3 + score4 + score5
print(total_score)

# 평균
avg_score = total_score / 5
print(avg_score)

# 연령대
age_group = age//10*10
print(age_group)

1104
220.8
20


[문2]  f-String을 사용해 위 변수들 값이 다음과 같은 형태로 출력되도록 코드를 작성하시오.

~~~
이름: 홍길동
연령대: 20대
점수: 153, 204, 259, 221, 267
합계: 1,104 / 평균: 220.80
~~~


In [ ]:
# 확인
print(f'이름: {name}')
print(f'연령대: {age_group}대')
print(f'점수: {score1}, {score2}, {score3}, {score4}, {score5}')
print(f'합계: {total_score:,} / 평균: {avg_score:.2f}')

## 2.6.문자열 관련 주요 메서드

* 변수(또는 개체) 뒤에 '.함수' 형태로 사용하는 함수를 **메서드(Method)** 라고 부른다.
- 아래 언급된 메서드 외에도 문자열 관련 메서드가 상당히 많으니 관련 자료를 참고하기 바람
- 문자열을 변환하는 메서드를 사용해도 그 결과가 변수에 반영되지는 않는다.

In [68]:
# 대문자로 변환
str = 'I have a Dream'
print(str.upper())
print(str)

I HAVE A DREAM
I have a Dream


- 변환 결과를 변수에 반영하려면 메소드가 반환하는 값을 변수에 대입해야 한다.

In [70]:
# 대문자로 변환
print(str.upper())

# 변수에 반영
str = str.upper()
print(str)

I HAVE A DREAM
I HAVE A DREAM


**1) 대/소문자 변환**

- **capitalize()** 메소드: 문자열 첫 문자만 대문자, 나머지 모두 소문자로 변환
- **title()** 메소드: 단어 첫 문자만 대문자, 나머지는 모두 소문자로 변환
- **upper()** 메소드: 모든 문자를 대문자로 변환
- **lower()** 메소드: 모든 문자를 소문자로 변환

In [73]:
# 대/소문자 변환
print(str.lower())
print(str.title())
print(str.capitalize())

i have a dream
I Have A Dream
I have a dream


**2) 자릿수 지정과 정렬**

- **rjust(x)** 메소드: 지정한 x 만큼의 자릿수를 확보한 후 오른쪽 정렬
- **ljust(x)** 메소드: 지정한 x 만큼의 자릿수를 확보한 후 왼쪽 정렬
- **center(x)** 메소드: 지정한 x 만큼의 자릿수를 확보한 후 가운데 정렬

In [75]:
# 지정한 길이의 문자열이 되도록 공백을 추가하고 정렬
str2 = 'Dream'

print('[' + str2.rjust(10)+']')
print('[' + str2.ljust(10)+']')
print('[' + str2.center(10)+']')

[     Dream]
[Dream     ]
[  Dream   ]


**3) 일부 변경 및 제거**

- **replace(x, y)** 메소드: 문자열에 포함된 일부 문자열 x를 문자열 y로 변경
- **strip(x)** 메소드: 문자열에서 양쪽의 공백 또는 특수 문자를 제거

In [76]:
# 문자열 변경('now' --> 'now ')
s = 'nowhere'
s.replace('now', 'now ')

'now here'

In [92]:
# 앞 뒤 빈 문자열, 특수 문자 제거
print('      world'.strip())
print('#world#$$#'.strip())
print('#world#$$#'.strip('$#'))

world
#world#$$#
world


- replace() 메소드를 사용하는 경우가 많으니 꼭 기억합시다.

In [86]:
# 문자열 변수
num = '123,456'

num = int(num.replace(',', ''))
type(num)

int

**<연습문제>**

[문1] 다음 코드의 실행 결과가 '파이썬을 열심히 공부할게요!'가 되게 수정해 보시오.



In [93]:
# 문자열 변수 선언
my_str = '파이썬을 대충 공부할게요!'

# 문자열 변환
my_str.replace('대충', '열심히')

'파이썬을 열심히 공부할게요!'

**4) 분리와 연결**

- **split((x)** 메소드: x 문자를 구분자로 하여 문자열을 분리하여 리스트로 변환
(공백이 기본 구분자)
- **x.join(s)** 메소드: x를 구분자로 하여 요소들을 연결한 문자열로 변환

In [94]:
str = 'I have a Dream!'

# 문자열 분리
str.split()

['I', 'have', 'a', 'Dream!']

In [96]:
str = '2020-09-14'

# 문자열 분리
print(str.split('-'))
print(str.split('-')[1])

['2020', '09', '14']
09


In [97]:
# 문자열 변수
str = '1/2/3/4/5/6/7/8/9'

# 문자열 분리
str_new = str.split('/')

# 확인
print(str_new)

['1', '2', '3', '4', '5', '6', '7', '8', '9']


In [98]:
# 문자열 변수
str = '2020-09-14'

# 문자열 분리
year = str.split('-')[0]
month = str.split('-')[1]
day = str.split('-')[2]

# 확인
print(f'오늘은 {year}년 {month}월 {day}일 입니다.')

오늘은 2020년 09월 14일 입니다.


- join() 메소드로 구분자를 두어 연결된 문자열을 만든다.

In [99]:
# 중간에 특정 문자열 삽입
str = 'PYTHON'
s1 = '/'.join(str)
print(s1)

P/Y/T/H/O/N


In [100]:
# 요소를 연결해서 문자열로
name = ['홍길동', '한사랑', '강우동', '박여인']
str = ','.join(name)

# 확인
str

'홍길동,한사랑,강우동,박여인'

In [101]:
# 문자열 str를 구분자로 나누어 리스트로
str3 = str.split(',')

# 확인
str3

['홍길동', '한사랑', '강우동', '박여인']

**<연습문제>**

[문1] 다음 주어진 문자열을 적절한 구분자를 기준으로 분리하시오.



In [102]:
# 문자열 변수
my_str = '15, 23, 52, 41, 25, 60, 47, 28, 49, 60'

# 문자열 분리
my_str.split(',')

['15', ' 23', ' 52', ' 41', ' 25', ' 60', ' 47', ' 28', ' 49', ' 60']

[문2] 다음 주어진 리스트의 요소를 yyyy-mm-dd 형태로 연결해 dt 변수에 담고 결과를  확인하시오.


In [104]:
# 리스트 만들기
my_list = ['2020', '09', '14']

# 문자열 연결
dt = '-'.join(my_list)

# 확인
print(dt)

2020-09-14


[문3] 위에서 배운 문자열 관련 메소드를 사용해 다음 주어진 리스트의 요소를 모두 대문자로 바꾸시오.


In [108]:
# 리스트 만들기
happy = ['h', 'a', 'p',  'p', 'y']

# 대문자로 변경
happy = ''.join(happy).upper()
print(happy)
print(type(happy))

# 확인
print(list(happy))

HAPPY
<class 'str'>
['H', 'A', 'P', 'P', 'Y']
